services/121-service/src/transaction-job-processors/transaction-job-processors.service.ts

```diff
@@ -217,30 +214,143 @@ export class TransactionJobProcessorsService {
+  ): Promise<void> {
+    // 1. Get additional data
+    const registration = await this.getRegistrationOrThrow(
+      transactionJob.referenceId,
+    );
+    const oldRegistration = structuredClone(registration);
+    const financialServiceProvider =
+      await this.getFinancialServiceProviderOrThrow(
+        FinancialServiceProviderName.safaricom,
+      );
+
+    // 2. Check if all required properties are present. If not, create a failed transaction and throw an error.
+    for (const [name, value] of Object.entries(transactionJob)) {
+      // Define "empty" based on your needs. Here, we check for null, undefined, or an empty string.
+      if (value === null || value === undefined || value === '') {
+        const errorText = `Property ${name} is undefined`;
+        await this.createTransactionAndUpdateRegistration({
+          programId: transactionJob.programId,
+          paymentNumber: transactionJob.paymentNumber,
+          userId: transactionJob.userId,
+          calculatedTransferAmountInMajorUnit: transactionJob.transactionAmount,
+          financialServiceProviderId: financialServiceProvider.id,
+          registration,
+          oldRegistration,
+          isRetry: transactionJob.isRetry,
+          status: TransactionStatusEnum.error,
+          errorText,
+        });
+        return;
+      }
+    }
+
+    // 3. Save the transaction into database with 'waiting' status and then call to safaricom for payouts. And then safaricom will return the
+    // actual payout status in callback url and then we will update the transaction status to success or error after procced the callback.
+    const transaction = await this.createTransactionAndUpdateRegistration({
+      programId: transactionJob.programId,
+      paymentNumber: transactionJob.paymentNumber,
+      userId: transactionJob.userId,
+      calculatedTransferAmountInMajorUnit: transactionJob.transactionAmount,
+      financialServiceProviderId: financialServiceProvider.id,
+      registration,
+      oldRegistration,
+      isRetry: transactionJob.isRetry,
+      status: TransactionStatusEnum.waiting, // This will only go to 'success' via callback
+    });
+
+    // 4. Start the transfer, if failure update to error transaction and return early
+    try {
+      await this.safaricomService.doTransfer({
+        transactionId: transaction.id,
+        transferAmount: transactionJob.transactionAmount,
+        phoneNumber: transactionJob.phoneNumber!,
+        idNumber: transactionJob.idNumber!,
+        remarks: `Payment ${transactionJob.paymentNumber}`,
+        originatorConversationId: transactionJob.originatorConversationId!,
+      });
+    } catch (error) {
+      if (error instanceof SafaricomApiError) {
+        await this.transactionScopedRepository.update(
+          { id: transaction.id },
+          { status: TransactionStatusEnum.error, errorMessage: error?.message },
+        );
+        return;
+      }
+      if (error instanceof QueryFailedError) {
```

Comment:
we are specifically catching the error in case of duplicate originatorConversationId here, because that will break on the unique constraint in our entity, so that is a QueryFailedError.